src/ScheduleFunctions.cpp

```diff
@@ -1927,9 +1932,24 @@ bool validate_schedule(Function f, Stmt s, const Target &target, bool is_output,
         if (s.allow_race_conditions()) {
             allow_race_conditions_count++;
         }
+        bool need_propagate_parallel = true;
+        while (need_propagate_parallel) {
```

Comment:
The splits are ordered, so one reverse pass through the list should be sufficient instead of iterating until convergence.

Also, there are four types of Split, and the concept of a child var varies across them. For a vanilla split, inner and outer are the children and old_var is the parent. For rename and purify, the outer is the child and the inner is meaningless. For fuse, old_var is the child and inner/outer are the parents. I acknowledge that this is gross and should be refactored.

I think the logic should be:

```
if (split.is_split() && (outer is parallel || inner is parallel)) insert old_var
else if (split.is_fuse() && old_var is parallel) insert inner and outer
else if ((split.is_rename() || split.is_purify()) && outer is parallel) insert outer
```

Reply:
> The splits are ordered

Good to know.

> I acknowledge that this is gross and should be refactored.

I am totally gonna capture that in a comment in the code.